### ISSUE TITLE
Two factor support

### ISSUE THREAD

**Issue by danschultzer** [danschultzer/pow#6]
Add two factor support to Pow with at least OTP and FIDO U2F support.

Good primer for ensuring secure and useful two factor setup: https://medium.com/@stuartschechter/before-you-turn-on-two-factor-authentication-27148cc5b9a1

**Comment by danschultzer** [comment]
A WebAuthn library that could be tested with: https://github.com/tanguilp/wax

**Comment by danschultzer** [comment]
WebAuthn standard has been finalized by W3C!

We should consider a pure single factor Webauthn registration/authentication process where user id and password is not required (or user id is the the webauthn credential id(s). On top of this there will also be two factor.

I do think it’s the future of web authentication, far more than provider login.

**Comment by danschultzer** [comment]
From [the elixirforum](https://elixirforum.com/t/pow-robust-modular-extendable-user-authentication-and-management-system/15807/82?u=danschultzer), there's a question on multiple user id field authentication:

> I’m just a bit curious. ^^
> Is it possible (or could it be) to pass a list of fields to `:user_id_field` , so that user can authenticate through the field of their choice?
> 
> ```
> defmodule MyApp.Users.User do
>   use Ecto.Schema
>   use Pow.Ecto.Schema,
>     user_id_field: [:email, :pseudo, :phone]
> 
>   # ...
> end
> ```

This may fit with the user id less WebAuthn setup.

**Comment by danhunsaker** [comment]
What's the feeling on using external libraries to make this go faster/smoother?

Specifically, [elixir2fa](https://github.com/codenoid/elixir-two-factor-auth) and [passwordless_auth](https://github.com/madebymany/passwordless_auth)?

**Comment by danschultzer** [comment]
I don't mind using a library to do the heavy lifting at all, but I don't plan to add any more dependencies to Pow so it would have to be an adapter setup, or an extension. No matter how it will be, I believe the biggest hurdle is to refactor the code so it's flexible enough to handle a [custom auth flow](https://github.com/danschultzer/pow/issues/234#issuecomment-513581445) (independent on what auth logic is used).

**Comment by danhunsaker** [comment]
An extension is probably sanest, then. Refactoring is certainly quite the task.

**Comment by danschultzer** [comment]
Yeah. I've been too busy, but now with [API integration guide](https://github.com/danschultzer/pow/pull/247) and [out-of-the-box distributed Mnesia](https://github.com/danschultzer/pow/pull/233) released soon, I believe this is probably the next thing I'll take a dig at 😄

**Comment by danhunsaker** [comment]
Sounds awesome! Looking forward to removing my custom-rolled 2FA stuff in favor of the official implementation. (Fewer things I can do wrong if I'm not the one doing them. :joy: )

**Comment by danschultzer** [comment]
For sure. If you can share the code, I would be happy to see how you solved it (maybe gist or repo)!

**Comment by danhunsaker** [comment]
Still implementing, but an extension that hooks into `before_respond/4` is my current plan.

**Comment by OldhamMade** [comment]
Any updates on how this is going? Would love to see MFA in Pow!
